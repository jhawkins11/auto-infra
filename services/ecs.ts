import { InitAnswers } from '../types/prompts.js';
import { ensureDirectoryExists, writeConfigFile } from './fileWriter.js';

const taskSizeToResources: Record<
  NonNullable<InitAnswers['ecsTaskSize']>,
  { cpu: number; memory: number }
> = {
  small: { cpu: 512, memory: 1024 },
  medium: { cpu: 1024, memory: 2048 },
  large: { cpu: 2048, memory: 4096 },
};

export const generateEcsConfig = async (
  answers: InitAnswers,
): Promise<void> => {
  const configDirectory = `./${answers.projectName}`;
  const ecsConfigPath = `${configDirectory}/ecs.tfvars`;

  const resources = taskSizeToResources[answers.ecsTaskSize ?? 'small'];

  const ecsConfig = {
    project_name: answers.projectName,
    aws_region: answers.awsRegion,
    container_image: answers.containerImageUrl ?? '',
    container_port: Number(answers.containerPort ?? 3000),
    task_cpu: resources.cpu,
    task_memory: resources.memory,
    desired_count: Number(answers.desiredCount ?? 2),
    health_check_path: answers.healthCheckPath ?? '/healthz',
  };

  await ensureDirectoryExists(ecsConfigPath);
  await writeConfigFile(ecsConfigPath, ecsConfig);
};
