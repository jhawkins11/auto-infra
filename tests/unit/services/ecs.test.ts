jest.mock('node:fs/promises', () => ({
  writeFile: jest.fn(),
  mkdir: jest.fn(),
}));

import * as fs from 'node:fs/promises';

import { generateEcsConfig } from '../../../services/ecs.js';
import { mustExist } from '../../utils/assertions.js';
import { createTestAnswers } from '../../utils/createTestAnswers.js';

describe('generateEcsConfig', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should write an ecs.tfvars file with the correct content', async () => {
    const answers = createTestAnswers({
      projectName: 'ecs-project',
      awsRegion: 'us-west-1',
      ecsTaskSize: 'medium',
      containerImageUrl: 'ghcr.io/example/app:1.2.3',
      containerPort: 4000,
      desiredCount: 4,
      healthCheckPath: '/hc',
    });

    await generateEcsConfig(answers);

    const mkdirMock = jest.mocked(fs.mkdir);
    const writeFileMock = jest.mocked(fs.writeFile);

    expect(mkdirMock).toHaveBeenCalledTimes(1);
    expect(mkdirMock).toHaveBeenCalledWith(`./${answers.projectName}`, {
      recursive: true,
    });

    const writeCalls = writeFileMock.mock.calls as Array<[string, string]>;
    const ecsCall = writeCalls.find((c) =>
      String(c[0]).endsWith('/ecs.tfvars'),
    );
    if (!ecsCall) {
      throw new Error('ecs.tfvars not written');
    }
    const [pathArgument, contentArgument] = ecsCall as [string, string];
    expect(pathArgument).toBe(`./${answers.projectName}/ecs.tfvars`);
    expect(contentArgument).toMatchSnapshot();
  });

  it('should use default values when optional fields are missing', async () => {
    const answers = createTestAnswers({ projectName: 'ecs-defaults' });
    // remove optional fields
    delete (answers as Partial<typeof answers>).ecsTaskSize;
    delete (answers as Partial<typeof answers>).containerPort;

    await generateEcsConfig(answers);

    const writeFileMock = jest.mocked(fs.writeFile);
    const writeCalls = writeFileMock.mock.calls as Array<[string, string]>;
    const ecsCall = mustExist(
      writeCalls.find((c) => String(c[0]).endsWith('/ecs.tfvars')),
    );
    const [, contentArgument] = ecsCall as [string, string];

    expect(String(contentArgument)).toContain('task_cpu');
    expect(String(contentArgument)).toContain('task_memory');
    expect(String(contentArgument)).toContain('container_port = 3000');
  });
});
