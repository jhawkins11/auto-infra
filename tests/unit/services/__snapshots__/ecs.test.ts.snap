// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`generateEcsConfig should write an ecs.tfvars file with the correct content 1`] = `
"project_name = "ecs-project"
aws_region = "us-west-1"
container_image = "ghcr.io/example/app:1.2.3"
container_port = 4000
task_cpu = 1024
task_memory = 2048
desired_count = 4
health_check_path = "/hc""
`;
